### PR TITLE
x64: Add small/build AMI/ASG

### PR DESCRIPTION
**Pull request by phlax**

### CODE REVIEW

File: instances/azp-small-build-asg/iam.tf [hunk 1]
```diff
@@ -0,0 +1,104 @@
+# Creates an IAM Role where instances themselves can then detach themselves from ASG.
+data "aws_iam_policy_document" "asg_detach_nano_instances" {
```
Comment (phlax) [inline instances/azp-small-build-asg/iam.tf]:
```suggestion
data "aws_iam_policy_document" "asg_detach_small_instances" {
```

File: instances/azp-small-build-asg/iam.tf [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+      "autoscaling:DetachInstances"
+    ]
+
+    resources = [
+      "arn:aws:autoscaling:*:${var.aws_account_id}:autoScalingGroup:*:autoScalingGroupName/${local.asg_name}",
+    ]
+  }
+}
+
+data "aws_iam_policy_document" "nano_instance_assume_role_policy" {
```
Comment (phlax) [inline instances/azp-small-build-asg/iam.tf]:
```suggestion
data "aws_iam_policy_document" "small_instance_assume_role_policy" {
```

File: ami-build/scripts/run-fun.sh [hunk 3]
```diff
@@ -187,8 +187,8 @@ _agent_start_finalize () {
 
 
 agent_start_build () {
-    local bazel_cache_bucket="$6" \
-          cache_prefix="$7" \
+    local bazel_cache_bucket="$5" \
+          cache_prefix="$6" \
```
Comment (mum4k) [inline ami-build/scripts/run-fun.sh:198]: (optional, pre-existing) This is a large number of arguments. Consider documenting them and/or giving them local names.

File: instances/azp-small-build-asg/init.sh.tpl [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+# shellcheck disable=SC2154
+agent_start_build \
+    "${azp_pool_name}" \
+    "${asg_name}" \
+    "${instance_profile_arn}" \
+    "${role_name}" \
+    ${CACHE_BUCKET} \
+    ${CACHE_PREFIX} \
+
+# figure out what the CACHE_ vars are
```
Comment (mum4k) [inline instances/azp-small-build-asg/init.sh.tpl]: Did you intend to remove this comment before we merge it in?

**Review by mum4k — DISMISSED** [review]

File: instances/azp-small-build-asg/init.sh.tpl (hunk 4 above)
Comment (phlax) [reply instances/azp-small-build-asg/init.sh.tpl]: ah yep - now i rem why it was WIP - this needs to be resolved i think

File: ami-build/scripts/run-fun.sh (hunk 3 above)
Comment (phlax) [reply ami-build/scripts/run-fun.sh:198]: yeah, this is a bit non-intuitive - it passes through the args and adds a couple

i believe that the correction i made here is necessary but i need to confirm that

**Comment by phlax** [comment]
~just realizing that this is also missing a packer .json file - will look now at adding~

nm - i think the hcl file should do it

File: ami-build/scripts/run-fun.sh (hunk 3 above)
Comment (phlax) [reply ami-build/scripts/run-fun.sh:198]: on further investigation these are the same names that are used in terraform - not sure how they could be named locally other than what is there now

Comment (phlax) [reply ami-build/scripts/run-fun.sh:198]: still not 100% about the args numbers but i think this is correct and not sure how to test without running it

Comment (mum4k) [reply ami-build/scripts/run-fun.sh:198]: Ack on running being the only known way to test this.

On the documentation - can we add a function doc for `agent_start_build` that would list the arguments this function takes and their meaning?

Comment (phlax) [reply ami-build/scripts/run-fun.sh:198]: added

**Review by mum4k — DISMISSED** [review]

**Review by mum4k — APPROVED** [review]

**Pull request merged**